blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
Let's take a break from calling the code of each other rather complicated and consider an example. 
Suppose instead of `Softmax` I want to normalize my scores in a smoother way, e.g. by applying sigmoids instead of exponential function like we did in the TIMIT paper. With the wrapper bricks I advocate one only has to implement the required logic for the simplest case when the input is a matrix. Reshapes required to make it work with more-dimensional inputs can be added by combining the new brick `SmoothSoftmax` with `WithExtraDims` just like it is done in this PR.

On the other hand, your `AxisSoftmax` provides no reusable code for that. One has to copy and modify it in good old Groundhog style.

Please tell me if there is a flaw in this reasoning.